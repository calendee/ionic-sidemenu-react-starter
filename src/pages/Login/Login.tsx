import React, { useContext, useState, useEffect } from 'react';
import {
  IonButton,
  IonContent,
  IonInput,
  IonItem,
  IonLabel,
  IonList,
  IonPage,
} from '@ionic/react';
import { RouteComponentProps } from 'react-router';

import { ActionTypes } from 'providers/Events/eventsActions';
import { EventsContext } from 'providers/Events/EventsContextProvider';
import { UserContext } from 'providers/User/UserContextProvider';
// @ts-ignore
import Modal from '../../components/Modal/Modal.tsx';
import useInput from '../../hooks/useInput';

interface LoginPageProps extends RouteComponentProps {}

const LoginPage: React.FC<LoginPageProps> = ({ history, location }) => {
  const [isOpen, setIsOpen] = useState(true);
  const [formValid, setFormValid] = useState(false);
  const [formSubmitted, setFormSubmitted] = useState(false);
  const { dispatch: eventsDispatch } = useContext(EventsContext);
  const { setUser } = useContext(UserContext);
  const [firstName, setFirstName] = useInput('');
  const [lastName, setLastName] = useInput('');

  /**
   * Validate the form
   */
  useEffect(() => {
    if (firstName && lastName) {
      setFormValid(true);
    } else {
      setFormValid(false);
    }
  }, [firstName, lastName, setFormValid]);

  const login = (e?: React.FormEvent | null) => {
    if (e) {
      e.preventDefault();
    }
    if (firstName && lastName) {
      setIsOpen(false);
      setTimeout(() => {
        setUser({
          firstName,
          lastName,
          uid: Math.random()
            .toString(36)
            .substring(7),
        });

        // Redirect authenticated user to home page
        if (location.pathname === '/login') {
          history.replace('/home');
        }
      });
    } else {
      setFormSubmitted(true);
      setFormValid(false);
    }
  };

  const recordEvent = (event: string) => {
    eventsDispatch({
      type: ActionTypes.RECORD_EVENT,
      payload: {
        event,
      },
    });
  };

  return (
    <IonPage data-testid="login-page">
      <IonContent>
        <Modal close={null} open={isOpen} title="Login">
          <IonContent>
            <form onSubmit={login}>
              <IonList>
                <IonItem>
                  <IonLabel>First Name</IonLabel>
                  <IonInput
                    data-testid="firstNameField"
                    id="firstName"
                    name="firstName"
                    onIonChange={setFirstName}
                    placeholder="Enter your first name"
                    value={firstName}
                  ></IonInput>
                </IonItem>
                <IonItem>
                  <IonLabel>Last Name</IonLabel>
                  <IonInput
                    id="lastName"
                    name="lastName"
                    onIonChange={setLastName}
                    placeholder="Enter your last name"
                    value={lastName}
                  ></IonInput>
                </IonItem>
              </IonList>
              <IonButton
                color="primary"
                // disabled={!formValid}
                expand="block"
                style={{ marginLeft: '10px' }}
                type="submit"
                onClick={() => {
                  recordEvent('login-button-clicked');
                  login();
                }}
              >
                Login
              </IonButton>
            </form>
            {formSubmitted && !formValid ? (
              <p>Please enter your first and last name</p>
            ) : null}
          </IonContent>
        </Modal>
      </IonContent>
    </IonPage>
  );
};

export default LoginPage;
