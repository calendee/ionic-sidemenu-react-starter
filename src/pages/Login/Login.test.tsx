import React from 'react';
import {
  fireEvent,
  renderWithUserAndEvents,
  wait,
} from '../../../test/test-utils';

import Login from './Login';

// TODO: Replace this mock with actual import
// Unable to do because get `customElements not defined`
jest.mock('../../components/Modal/Modal', () => {
  const mockModal = ({ children }: { children: any }) => (
    <div id="fake-modal">{children}</div>
  );
  return mockModal;
});

describe('Routes', () => {
  test('renders login modal with no user', async () => {
    // @ts-ignore
    const { container } = renderWithUserAndEvents(<Login />);
    await wait();
    expect(container).toMatchSnapshot();
  });

  test('renders login error', async () => {
    // @ts-ignore
    const { container, getByText } = renderWithUserAndEvents(<Login />);
    const loginButton = getByText(/login/i, {
      selector: 'ion-button',
    });

    // Test failing to login
    fireEvent.click(loginButton);
    await wait();
    expect(container).toMatchSnapshot();
  });
});
