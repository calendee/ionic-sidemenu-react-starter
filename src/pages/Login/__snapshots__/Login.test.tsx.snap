// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes renders login error 1`] = `
<div>
  <ion-app>
    <div
      class="ion-page"
      data-testid="login-page"
    >
      <ion-content>
        <div
          id="fake-modal"
        >
          <ion-content>
            <form>
              <ion-list>
                <ion-item>
                  <ion-label>
                    First Name
                  </ion-label>
                  <ion-input
                    data-testid="firstNameField"
                    id="firstName"
                    name="firstName"
                    placeholder="Enter your first name"
                    value=""
                  />
                </ion-item>
                <ion-item>
                  <ion-label>
                    Last Name
                  </ion-label>
                  <ion-input
                    data-testid="firstNameField"
                    id="lastName"
                    name="lastName"
                    placeholder="Enter your last name"
                    value=""
                  />
                </ion-item>
              </ion-list>
              <ion-button
                color="primary"
                expand="block"
                style="margin-left: 10px;"
                type="submit"
              >
                Login
              </ion-button>
            </form>
            <p>
              Please enter your first and last name
            </p>
          </ion-content>
        </div>
      </ion-content>
    </div>
  </ion-app>
</div>
`;

exports[`Routes renders login modal with no user 1`] = `
<div>
  <ion-app>
    <div
      class="ion-page"
      data-testid="login-page"
    >
      <ion-content>
        <div
          id="fake-modal"
        >
          <ion-content>
            <form>
              <ion-list>
                <ion-item>
                  <ion-label>
                    First Name
                  </ion-label>
                  <ion-input
                    data-testid="firstNameField"
                    id="firstName"
                    name="firstName"
                    placeholder="Enter your first name"
                    value=""
                  />
                </ion-item>
                <ion-item>
                  <ion-label>
                    Last Name
                  </ion-label>
                  <ion-input
                    data-testid="firstNameField"
                    id="lastName"
                    name="lastName"
                    placeholder="Enter your last name"
                    value=""
                  />
                </ion-item>
              </ion-list>
              <ion-button
                color="primary"
                expand="block"
                style="margin-left: 10px;"
                type="submit"
              >
                Login
              </ion-button>
            </form>
          </ion-content>
        </div>
      </ion-content>
    </div>
  </ion-app>
</div>
`;
